### PR TITLE
Remove deprecated usage of `.dtype()` method

### PR DESCRIPTION
**Pull request by JanuszL**
- adds to_numpy_type method to convert between DALI
  and NumPy types
- adjusts DALI plugins and tests to not use `dtype()`

Signed-off-by: Janusz Lisiecki <jlisiecki@nvidia.com>

## Category:

**Other** (*e.g. Documentation, Tests, Configuration*)

## Description:
- adds to_numpy_type method to convert between DALI
  and NumPy types
- adjusts DALI plugins and tests to not use `dtype()`

### Affected modules and functionalities:
- plugin/mxnet.py
- plugin/paddle.py
- plugin/pytorch.py
- types.py
- test_backend_impl.py
- test_backend_impl_gpu.py
- test_utils.py
- util/pybind.h

### Key points relevant for the review:
- if there is any place that have been missed

### Tests
- [X] Existing tests apply
- [ ] New tests added
  - [ ] Python tests
  - [ ] GTests
  - [ ] Benchmark
  - [ ] Other
- [ ] N/A

### Documentation
- [ ] Existing documentation applies
- [ ] Documentation updated
  - [X] Docstring
  - [ ] Doxygen
  - [ ] RST
  - [ ] Jupyter
  - [ ] Other
- [ ] N/A

#### Requirements
- [ ] Implements new requirements
- [ ] Affects existing requirements
- [X] N/A

**REQ IDs**: N/A

**JIRA TASK**: N/A

### CODE REVIEW

**Comment by JanuszL** [comment]
!build

File: dali/test/python/test_backend_impl.py [hunk 1]
```diff
@@ -236,7 +236,7 @@ def dtype_pipeline(np_type, placeholder_dali_type):
 
 
 def test_dtype_converion():
-    dali_types = [types._DALIDataType.INT8, types._DALIDataType.UINT64, types._DALIDataType.FLOAT16]
```
Comment (klecki) [inline dali/test/python/test_backend_impl.py:239]: This was intentional, please revert.

We should probably add a comment why. It already is explained a bit in backend_impl though - the _DALIDataType is the type inheriting from DALIDataType, extended with `__call__` that supports returning the numpy/array_interface type string.

We want to make sure, that it works as a `dtype` parameter to operators without any issues, as someone may use it that way.

**Review by klecki — CHANGES_REQUESTED** [review]

File: dali/python/nvidia/dali/types.py [hunk 2]
```diff
@@ -104,6 +104,30 @@ def _default_converter(dtype, default_value):
...
+        DALIDataType.UINT8   : np.uint8,
+        DALIDataType.UINT16  : np.uint16,
+        DALIDataType.UINT32  : np.uint32,
+        DALIDataType.UINT64  : np.uint64,
+        DALIDataType.INT8    : np.int8,
+        DALIDataType.INT16   : np.int16,
+        DALIDataType.INT32   : np.int32,
+        DALIDataType.INT64   : np.int64,
+        DALIDataType.FLOAT16 : np.float16,
+        DALIDataType.FLOAT   : np.float32,
```
[the remark this answers is not in the document]
Comment (JanuszL) [reply dali/python/nvidia/dali/types.py:126]: I can use `np.float_` instead if you like.

File: dali/test/python/test_backend_impl.py (hunk 1 above)
Comment (JanuszL) [reply dali/test/python/test_backend_impl.py:239]: Fixed. nvidia.dali.backend_impl.types collided with nvidia.dali.types

File: dali/python/nvidia/dali/plugin/pytorch.py [hunk 3]
```diff
@@ -52,10 +51,9 @@ def feed_ndarray(dali_tensor, arr, cuda_stream = None):
                     if we are copying to a tensor allocated with torch.zeros(...))
     """
     if isinstance(dali_tensor, (TensorListCPU, TensorListGPU)):
-        dali_type = dali_tensor[0].dtype()
+        dali_type = dali_tensor[0].dtype
     else:
-        dali_type = dali_tensor.dtype()
-    dali_type = np.dtype(dali_type)
+        dali_type = dali_tensor.dtype
```
[the remark this answers is not in the document]
Comment (JanuszL) [reply dali/python/nvidia/dali/plugin/pytorch.py]: Fixed

File: dali/python/nvidia/dali/plugin/pytorch.py [hunk 4]
```diff
@@ -21,18 +21,17 @@
...
 to_torch_type = {
-    np.dtype(np.float32) : torch.float32,
-    np.dtype(np.float64) : torch.float64,
-    np.dtype(np.float16) : torch.float16,
-    np.dtype(np.uint8)   : torch.uint8,
-    np.dtype(np.int8)    : torch.int8,
-    np.dtype(np.int16)   : torch.int16,
-    np.dtype(np.int32)   : torch.int32,
-    np.dtype(np.int64)   : torch.int64
+    types.FLOAT   : torch.float32,
```
[the remark this answers is not in the document]
Comment (JanuszL) [reply dali/python/nvidia/dali/plugin/pytorch.py]: Fixed

File: dali/python/nvidia/dali/plugin/paddle.py [hunk 5]
```diff
@@ -51,10 +50,6 @@ def to_paddle_type(tensor):
     if isinstance(tensor, (TensorListCPU, TensorListGPU)):
         tensor = tensor.at(0)
     dtype = tensor.dtype
```
[the remark this answers is not in the document]
Comment (JanuszL) [reply dali/python/nvidia/dali/plugin/paddle.py]: Fixed

File: dali/python/nvidia/dali/plugin/paddle.py [hunk 6]
```diff
@@ -26,16 +26,15 @@
...
-    "=e": fluid.core.VarDesc.VarType.FP16,
-    "=f": fluid.core.VarDesc.VarType.FP32,
-    "=d": fluid.core.VarDesc.VarType.FP64,
-    "=B": fluid.core.VarDesc.VarType.UINT8,
-    "=b": fluid.core.VarDesc.VarType.INT8,
-    "=h": fluid.core.VarDesc.VarType.INT16,
-    "=i": fluid.core.VarDesc.VarType.INT32,
-    "=q": fluid.core.VarDesc.VarType.INT64,
-    "=l": fluid.core.VarDesc.VarType.INT64
+    types.BOOL    : fluid.core.VarDesc.VarType.BOOL,
```
[the remark this answers is not in the document]
Comment (JanuszL) [reply dali/python/nvidia/dali/plugin/paddle.py]: Fixed

File: dali/python/nvidia/dali/plugin/mxnet.py [hunk 7]
```diff
@@ -53,10 +52,10 @@ def feed_ndarray(dali_tensor, arr, cuda_stream = None):
                     is expected.
     """
     if isinstance(dali_tensor, (TensorListCPU, TensorListGPU)):
-        dali_type = dali_tensor[0].dtype()
+        dali_type = dali_tensor[0].dtype
```
[the remark this answers is not in the document]
Comment (JanuszL) [reply dali/python/nvidia/dali/plugin/mxnet.py]: Fixed

File: dali/python/nvidia/dali/types.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (klecki) [reply dali/python/nvidia/dali/types.py:126]: No, I'm just sad about DALI.Float, let's keep it like this.

File: dali/python/nvidia/dali/types.py [hunk 8]
```diff
@@ -104,6 +104,30 @@ def _default_converter(dtype, default_value):
     else:
         return repr(_type_convert_value(dtype, default_value))
 
+"""
+Convest DALIDataType to numpy type
```
Comment (klecki) [inline dali/python/nvidia/dali/types.py:108]:
```suggestion
Convert DALIDataType to NumPy type
```

**Comment by JanuszL** [comment]
!build

**Review by klecki — APPROVED** [review]

File: dali/python/nvidia/dali/types.py (hunk 8 above)
Comment (JanuszL) [reply dali/python/nvidia/dali/types.py]: Fixed

File: dali/python/nvidia/dali/types.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (JanuszL) [reply dali/python/nvidia/dali/types.py]: Maybe we should have an alias.

**Comment by JanuszL** [comment]
!build

**Comment by JanuszL** [comment]
!build

**Review by jantonguirao — APPROVED** [review]

**Pull request merged**